src/resolver.js

```diff
@@ -0,0 +1,162 @@
+const punycode = require("punycode");
+const NodeCache = require("node-cache");
+const { NodeClient } = require("hs-client");
+
+const logger = require("./logger");
+
+// Match both `sia://HASH` and `HASH` links.
+const SKYLINK_REGEXP = /^(sia:\/\/)?[a-zA-Z0-9_-]{46}$/;
+const REGISTRY_ENTRY_REGEXP = /^skyns:\/\/(?<publickey>[a-zA-Z0-9%]+)\/(?<datakey>[a-zA-Z0-9%]+)$/;
+
+class ResolutionError extends Error {}
+
+/**
+ * @typedef {Object} HSRecord
+ * This documentation only mentiones properties used by this service. If you need more information,
+ * visit https://hsd-dev.org/guides/resource-records.html.
+ *
+ * @property {string}   type  Record type. In case of this application, we only care about TXT records.
+ * @property {string[]} [txt] Values for TXT records.
```

Comment:
> It's odd that the name is [txt]. Is that a convention for naming arrays?

No, it's JSDoc for a field that's optional. It may not exist on the returned object.

> Also, all of these type definitions are not really attached to anything, are they? Are they just instructions for the TS compiler? If that's the case shouldn't they be in a separate file, so the developer doesn't need to see them?

It's not attached to anything, because in this case they document a response from the HSD node. Not sure if I like creating a file with just JSDocs, though -- I added them here, as `Resolver` class is the only one that touches the HSD node.